File: Data_Structure_Go/Queue/LinkQueue.go
package main

import (
	"fmt"
	"sync"
)

// 使用链表实现队列
type LinkQueue struct {
	front *LinkNode //队头指针
	rear  *LinkNode //队尾指针
	size  int       //队列大小
	lock  sync.Mutex
}

// 链表节点
type LinkNode struct {
	Next  *LinkNode
	Value string
}

// 初始化队列
func (q *LinkQueue) Init() {
	q.front = nil
	q.rear = nil
	q.size = 0
}

// 入队
func (q *LinkQueue) Enqueue(value string) {
	q.lock.Lock()
	defer q.lock.Unlock()

	//创建新加入的节点
	NewNode := &LinkNode{Value: value}
	NewNode.Next = nil

	//判断队列是否为空
	if q.front == nil && q.rear == nil {
		q.front = NewNode
		q.rear = NewNode

	} else {
		q.rear.Next = NewNode
		q.rear = q.rear.Next
	}
	q.size++
}

// 出队
func (q *LinkQueue) Dequeue() string {
	q.lock.Lock()
	defer q.lock.Unlock()

	//判断队列是否为空
	if q.size == 0 {
		panic("Queue is empty")
	} else {
		value := q.front.Value
		q.front = q.front.Next
		q.size--
		//在golang语言中不需要手动释放节点，如果没有再被使用了，系统会自动释放

		if q.size == 0 {
			q.front = nil
			q.rear = nil
		}
		return value
	}
}

// 打印函数
func (q *LinkQueue) Show() {
	for n := q.front; n != nil; n = n.Next {
		fmt.Println(n.Value)
	}
}

func test2() {
	var q LinkQueue
	q.Init()
	q.Enqueue("a")
	q.Enqueue("b")
	q.Enqueue("c")
	q.Enqueue("d")
	//q.Show()

	q.Dequeue()
	q.Dequeue()
	q.Show()

}
func main3() {
	test2()
}

File: Data_Structure_Cpp/Queue/LinkQueue.cpp
//
// Created by huang on 2024/6/12.
//

#include "iostream"

using namespace std;

//定义链表节点
template<class T>
struct Node{
    T data;
    Node* Next;
};

//链表队列
template<class T>
class Queue{
public:
    //构造函数
    Queue():front(NULL),rear(NULL){}

    //入队
    void Enqueue(T v);

    //出队
    bool Dequeue(T &e);

    //检查队列是否为空
    bool IsEmpty();

    //获取队列大小
    int Size();

    //打印队列
    void Show();

    //析构函数释放空间
    ~Queue();

private:
    Node<T> *front;   //队头指针
    Node<T> *rear;    //队尾指针
    int len;          //兑长
};

//入队
template<class T>
void Queue<T>::Enqueue(T v){
    Node<T> *newNode = new Node<T>;
    newNode->data = v;
    newNode->Next = NULL;
    if (rear){
        rear->Next = newNode;  //对尾入队
        rear = newNode;
    } else
        front = rear = newNode;  //链表为空
    this->len++;
}

//出队
template<class T>
bool Queue<T>::Dequeue(T &e){
    if (IsEmpty())
        return false;
    Node<T> *temp = front;  //保存头节点防止锻链
    e = front->data;
    front = front->Next;
    this->len--;
    return true;
}

//检查队列是否为空
template<class T>
bool Queue<T>::IsEmpty(){
    if (this->front == NULL && this->front == this->rear)
        return true;
    return false;
}

//获取队列大小
template<class T>
int Queue<T>::Size(){
    return this->len;
}

//打印队列
template<class T>
void Queue<T>::Show(){
    if (IsEmpty()) {
        cout << "Queue is empty" << endl;
        return;
    }
    cout << "--------------" << endl;
    Node<T>* temp = front;
    while (temp) {
        cout << temp->data << " ";
        temp = temp->Next;
    }
    cout << endl << "--------------" << endl;
}

//析构函数释放空间
template<class T>
Queue<T>::~Queue(){
    // 遍历链表并逐个删除节点
    Node<T>* current = front;
    while (current != nullptr) {
        Node<T>* temp = current;
        current = current->Next;
        delete temp;
    }
    front = rear = nullptr; // 将头尾指针置空
}

int main() {
    Queue<int> q;
    int e;

    // q的入队
    q.Enqueue(1);
    q.Enqueue(2);
    q.Enqueue(3);
    q.Enqueue(4);
    q.Show();
    q.Dequeue(e);
    cout << e << endl;
    q.Show();

    return 0;
}